File: internal_testsuites/typescript/src/testsuite/duplicate_files_artifact_mount_test/duplicate_files_artifact_mount.test.ts
import {
    ContainerConfig,
    ContainerConfigBuilder,
    FilesArtifactUUID,
    PortProtocol,
    PortSpec,
    ServiceID
} from "kurtosis-sdk";
import {createEnclave} from "../../test_helpers/enclave_setup";
import log from "loglevel";
import {err, ok, Result} from "neverthrow";
import axios from "axios";

const TEST_NAME = "duplicate-files-artifact-mount"
const IS_PARTITIONING_ENABLED = false

const IMAGE = "flashspys/nginx-static"
const SERVICE_ID: ServiceID = "file-server"

const TEST_FILES_ARTIFACT_URL = "https://kurtosis-public-access.s3.us-east-1.amazonaws.com/test-artifacts/static-fileserver-files.tgz"

const USER_SERVICE_MOUNTPOINT_FOR_TEST_FILESARTIFACT  = "/static"

const DUPLICATE_MOUNTPOINT_DOCKER_DAEMON_ERR_MSG  = "Duplicate mount point"

const DUPLICATE_MOUNTPOINT_KUBERNETES_ERR_MSG = "mountPath: Invalid value: \"/static\": must be unique"

jest.setTimeout(180000)

test("Test two files artifacts mounted to the same location", async () => {
    // ------------------------------------- ENGINE SETUP ----------------------------------------------
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)

    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }

    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {

        // ------------------------------------- TEST SETUP ----------------------------------------------
        const storeFirstArtifactResult = await enclaveContext.storeWebFiles(TEST_FILES_ARTIFACT_URL)
        if (storeFirstArtifactResult.isErr()) { throw storeFirstArtifactResult.error }
        const firstFilesArtifactUuid = storeFirstArtifactResult.value;
        const storeSecondArtifactResult = await enclaveContext.storeWebFiles(TEST_FILES_ARTIFACT_URL)
        if (storeSecondArtifactResult.isErr()) { throw storeSecondArtifactResult.error }
        const secondFilesArtifactUuid = storeSecondArtifactResult.value;

        const filesArtifactMountpoints = new Map<FilesArtifactUUID, string>()
        filesArtifactMountpoints.set(firstFilesArtifactUuid, USER_SERVICE_MOUNTPOINT_FOR_TEST_FILESARTIFACT);
        filesArtifactMountpoints.set(secondFilesArtifactUuid, USER_SERVICE_MOUNTPOINT_FOR_TEST_FILESARTIFACT);
        const fileServerContainerConfig = getFileServerContainerConfig(filesArtifactMountpoints)


        // ------------------------------------- TEST RUN ----------------------------------------------
        const addServiceResult = await enclaveContext.addService(SERVICE_ID, fileServerContainerConfig)
        if(addServiceResult.isOk()){
            throw new Error(`Adding service '${SERVICE_ID}' should have failed and did not, because duplicated files artifact mountpoints ` +
                `'${filesArtifactMountpoints}' should throw an error`)
        }
        const errMsg = addServiceResult.error.message
        if(!errMsg.includes(DUPLICATE_MOUNTPOINT_DOCKER_DAEMON_ERR_MSG) && !errMsg.includes(DUPLICATE_MOUNTPOINT_KUBERNETES_ERR_MSG)){
            throw new Error(`Adding service "${SERVICE_ID}" has failed, but the error is not the duplicated-files-artifact-mountpoints-error that we expected, this is throwing this error instead:\n "${errMsg}"`)
        }
    }finally{
        stopEnclaveFunction()
    }
    jest.clearAllTimers()
})

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================

function getFileServerContainerConfig(filesArtifactMountpoints: Map<FilesArtifactUUID, string>): ContainerConfig {
    const containerConfig = new ContainerConfigBuilder(IMAGE)
        .withFiles(filesArtifactMountpoints)
        .build()

   return containerConfig
}

File: internal_testsuites/golang/testsuite/duplicate_files_artifact_mount_test/duplicate_files_artifact_mount_test.go
package duplicate_files_artifact_mount_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/kurtosis-tech/kurtosis/api/golang/core/lib/services"
	"github.com/stretchr/testify/require"
	"strings"
	"testing"
)

const (
	testName              = "duplicate-files-artifact-mount"
	isPartitioningEnabled = false

	image                        = "flashspys/nginx-static"
	serviceId services.ServiceID = "file-server"

	testFilesArtifactUrl = "https://kurtosis-public-access.s3.us-east-1.amazonaws.com/test-artifacts/static-fileserver-files.tgz"

	userServiceMountPointForTestFilesArtifact = "/static"

	duplicateMountpointKubernetesErrMsgSentence   = "Invalid value: \"/static\": must be unique"
	duplicateMountpointDockerDaemonErrMsgSentence = "Duplicate mount point"
)

func TestStoreWebFiles(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, stopEnclaveFunc, _, err := test_helpers.CreateEnclave(t, ctx, testName, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer stopEnclaveFunc()

	// ------------------------------------- TEST SETUP ----------------------------------------------
	firstFilesArtifactUuid, err := enclaveCtx.StoreWebFiles(context.Background(), testFilesArtifactUrl)
	require.NoError(t, err, "An error occurred storing the first files artifact")
	secondFilesArtifactUuid, err := enclaveCtx.StoreWebFiles(context.Background(), testFilesArtifactUrl)
	require.NoError(t, err, "An error occurred storing the second files artifact")

	filesArtifactMountpoints := map[services.FilesArtifactUUID]string{
		firstFilesArtifactUuid:  userServiceMountPointForTestFilesArtifact,
		secondFilesArtifactUuid: userServiceMountPointForTestFilesArtifact,
	}
	fileServerContainerConfig := getFileServerContainerConfig(filesArtifactMountpoints)

	// ------------------------------------- TEST RUN ----------------------------------------------
	_, err = enclaveCtx.AddService(serviceId, fileServerContainerConfig)
	require.Errorf(
		t,
		err,
		"Adding service '%v' should have failed and did not, because duplicated files artifact mountpoints "+
			"'%v' should throw an error",
		serviceId,
		filesArtifactMountpoints,
	)
	isExpectedErrorMsg := strings.Contains(err.Error(), duplicateMountpointKubernetesErrMsgSentence) || strings.Contains(err.Error(), duplicateMountpointDockerDaemonErrMsgSentence)

	require.True(t, isExpectedErrorMsg, "Adding service '%v' has failed, but the error is not the duplicated-files-artifact-mountpoints error "+
		"that we expected, this is throwing this error instead:\n%v",
		serviceId,
		err.Error(),
	)
}

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================
func getFileServerContainerConfig(filesArtifactMountpoints map[services.FilesArtifactUUID]string) *services.ContainerConfig {
	containerConfig := services.NewContainerConfigBuilder(
		image,
	).WithFiles(
		filesArtifactMountpoints,
	).Build()
	return containerConfig
}
